File: mobile/lib/models/shop.dart
class Location {
  final double latitude;
  final double longitude;

  Location({required this.latitude, required this.longitude});

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
    );
  }
}

class Shop {
  final String id;
  final String name;
  final List<Location> locations;
  final String image;

  Shop({
    required this.id,
    required this.name,
    required this.locations,
    required this.image,
  });

  factory Shop.fromJson(Map<String, dynamic> json) {
    return Shop(
      id: json['id'] as String,
      name: json['name'] as String,
      locations:
          (json['locations'] as List<dynamic>?)
              ?.map((e) => Location.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      image: json['image'] as String,
    );
  }
}

File: admin/hooks/shop.ts
import { useState, useEffect, useCallback } from "react";
import { auth } from "@/lib/firebase";

export type Location = {
  latitude: number;
  longitude: number;
};

export type Shop = {
  id: string;
  name: string;
  locations: Location[];
  image: string;
};

const useShops = () => {
  const [shops, setShops] = useState<Shop[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);

  const fetchShops = useCallback(async () => {
    setLoading(true);
    setError(null);

    try {
      const response = await fetch(`${process.env.NEXT_PUBLIC_BACKEND_URL}/listShop`);
      if (!response.ok) {
        throw new Error(`Error: ${response.status} - ${response.statusText}`);
      }
      const data = (await response.json()) satisfies Shop[];
      setShops(data);
    } catch (err) {
      console.error("Error fetching shops:", err);
      setError(err as Error);
    } finally {
      setLoading(false);
    }
  }, []);

  const deleteShop = async (id: string) => {
    try {
      const token = await auth.currentUser?.getIdToken();
      const response = await fetch(
        `${process.env.NEXT_PUBLIC_BACKEND_URL}/deleteShop`,
        {
          method: "DELETE",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
          body: JSON.stringify({ id }),
        }
      );

      if (response.ok) {
        await fetchShops();
        return true;
      }
      return false;
    } catch (err) {
      console.error("Delete shop failed:", err);
      return false;
    }
  };

  useEffect(() => {
    const load = async () => {
      await fetchShops();
    };
    load();
  }, [fetchShops]);

  return { shops, loading, error, refetch: fetchShops, deleteShop };
};

export default useShops;

